File: frontend/_test_result_card_geometry.js
/**
 * Result card geometry unit tests.
 *
 * Tests the pure function calcResultCardPosition() from _result_card_geometry.js.
 * These mirror the scenarios from ROUND9_LONG_TASK.md §Result Card Geometry.
 *
 * Run: node frontend/_test_result_card_geometry.js
 */
'use strict';
const { calcResultCardPosition, CARD_WIDTH, CARD_GAP } = require('./_result_card_geometry');

const failures = [];
function assert(label, condition, detail) {
  if (!condition) failures.push(`FAIL: ${label}${detail ? ' — ' + detail : ''}`);
  else console.log(`PASS: ${label}`);
}

// ── 1: Card width constant ────────────────────────
assert('CARD_WIDTH === 360', CARD_WIDTH === 360);
assert('CARD_GAP === 14', CARD_GAP === 14);

// ── 2: Position above float bar (fallback: floatWinBounds) ──
{
  const result = calcResultCardPosition(
    200,
    { x: 0, y: 0, width: 1920, height: 1080 },
    { x: 710, y: 1046, width: 500, height: 500 },
    null
  );
  // card bottom = (1046+500-34) - 14 = 1512 - 14 = 1498
  // card top = 1498 - 200 = 1298 — but workArea height is 1080, so clamp to 1080-200=880
  // Wait - let me recalculate: bar top = 1046 + 500 - 34 = 1512
  // card_bottom = 1512 - 14 = 1498 — but workArea bottom is 1080, so card_y clamped to 1080-200=880
  const expectedY = 1080 - 200; // clamped to workArea bottom - cardHeight
  assert('float bar fallback: card above bar',
    result.y >= 0 && result.y + 200 <= 1080,
    `y=${result.y}`);

  // card left = 710 + floor(500/2) - floor(360/2) = 710 + 250 - 180 = 780
  assert('float bar fallback: horizontal center',
    result.x === 780, `got x=${result.x}`);
}

// ── 3: Clamp to workArea left edge ───────────────
{
  // Float win positioned such that card would go left of workArea
  const result = calcResultCardPosition(
    200,
    { x: 0, y: 0, width: 1920, height: 1080 },
    null, null
  );
  // Fallback: centered on workArea horizontally
  assert('fallback: card left >= workArea.x',
    result.x >= 0, `x=${result.x}`);
  assert('fallback: card right <= workArea.width',
    result.x + CARD_WIDTH <= 1920, `right=${result.x + CARD_WIDTH}`);
}

// ── 4: Element positions (viewport→screen coord conversion) ─
{
  // Float window at screen (500, 200) with 500x500
  // Element position (viewport-relative): { left: 207, top: 466, right: 293, bottom: 500 }
  const floatWin = { x: 500, y: 200, width: 500, height: 500 };
  const eps = [{ left: 207, top: 466, right: 293, bottom: 500 }];
  const result = calcResultCardPosition(200, { x: 0, y: 0, width: 1920, height: 1080 }, floatWin, eps);

  // anchorTop = 200 + 466 = 666
  // anchorLeft = 500 + 207 = 707
  // anchorWidth = 293 - 207 = 86
  // cardX = 707 + 43 - 180 = 570
  // cardBottom = 666 - 14 = 652
  // cardY = 652 - 200 = 452
  assert('element positions: correct x',
    result.x === 570, `got x=${result.x}`);
  assert('element positions: correct y',
    result.y === 452, `got y=${result.y}`);
}

// ── 5: Clamp to workArea right edge ──────────────
{
  // Float window near right edge of 1920-px display
  // Bar center at x=1800 → card_x = 1800 - 180 = 1620 → right=1620+360=1980 > 1920
  const result = calcResultCardPosition(
    200,
    { x: 0, y: 0, width: 1920, height: 1080 },
    { x: 1550 /* bar center ≈ 1550+250=1800 */, y: 1000, width: 500, height: 500 },
    null
  );
  assert('clamp right: card right <= workArea.width',
    result.x + CARD_WIDTH <= 1920, `right=${result.x + CARD_WIDTH}`);
  assert('clamp right: card left >= 1560 (max left)',
    result.x >= 1560, `x=${result.x}`);
}

// ── 6: Clamp to workArea top edge ────────────────
{
  // Very high float window → card would go above workArea top
  const result = calcResultCardPosition(
    260,
    { x: 0, y: 0, width: 1920, height: 1080 },
    { x: 710, y: 50, width: 500, height: 500 },
    null
  );
  assert('clamp top: card top >= 0',
    result.y >= 0, `y=${result.y}`);
}

// ── 7: Last resort (no floatWin, no elementPositions) ──
{
  const result = calcResultCardPosition(
    200,
    { x: 0, y: 0, width: 1920, height: 1080 },
    null, null
  );
  // Center on primary display bottom: x = (1920 - 360)/2 = 780
  // y = 0 + 1080 - 200 - 14 - 34 = 832
  assert('last resort: center bottom',
    result.x === 780, `x=${result.x}`);
  assert('last resort: bottom position',
    result.y === 832, `y=${result.y}`);
}

// ── 8: Multi-display (secondary monitor) ─────────
{
  const result = calcResultCardPosition(
    200,
    { x: 1920, y: 0, width: 1920, height: 1080 },
    { x: 2630, y: 1046, width: 500, height: 500 },
    null
  );
  // bar center = 2630 + 250 = 2880
  // card_x = 2880 - 180 = 2700
  // card_x in secondary workArea [1920, 3840] ✓
  assert('multi-display: x within secondary workArea',
    result.x >= 1920 && result.x + CARD_WIDTH <= 3840,
    `x=${result.x}`);
  // bar_top = 1046 + 500 - 34 = 1512. card_bottom = 1512-14=1498 → clamp to 1080-200=880
  assert('multi-display: y within secondary workArea',
    result.y >= 0 && result.y + 200 <= 1080,
    `y=${result.y}`);
}

// ── 9: Card height extremes (150 min, 260 max) ───
{
  const resultMin = calcResultCardPosition(150, { x: 0, y: 0, width: 1920, height: 1080 }, null, null);
  const resultMax = calcResultCardPosition(260, { x: 0, y: 0, width: 1920, height: 1080 }, null, null);
  assert('min card height: y is valid',
    resultMin.y >= 0 && resultMin.y + 150 <= 1080,
    `y=${resultMin.y}`);
  assert('max card height: y is valid',
    resultMax.y >= 0 && resultMax.y + 260 <= 1080,
    `y=${resultMax.y}`);
  // Taller card means smaller y
  assert('taller card placed higher',
    resultMax.y < resultMin.y,
    `maxY=${resultMax.y} minY=${resultMin.y}`);
}

if (failures.length) {
  console.error(`\n--- ${failures.length} GEOMETRY FAILURE(S) ---`);
  for (const f of failures) console.error(f);
  process.exit(1);
}
console.log('\n--- ALL GEOMETRY TESTS PASSED ---');
process.exit(0);

File: tests/test_result_card_geometry.py
"""Phase 1 tests: result card geometry (size + position).

Verifies:
- Card width is 360px (per product requirement)
- Height is dynamic (150-240px range, max 260px)
- Position is above float bar with 12-16px gap
- Horizontal centering on float bar / element positions
- Clamping to display workArea
- Multi-monitor awareness
"""

from __future__ import annotations

import unittest
from unittest.mock import MagicMock


class ResultCardGeometryTests(unittest.TestCase):
    """Pure function tests for result card geometry calculations."""

    def _make_display(self, x=0, y=0, w=1920, h=1080):
        """Create a mock display object matching Electron's screen API."""
        d = MagicMock()
        d.workArea = {"x": x, "y": y, "width": w, "height": h}
        d.bounds = {"x": x, "y": y, "width": w, "height": h}
        return d

    def _make_element_position(self, left=0, top=0, right=500, bottom=500):
        """Create a mock element position dict."""
        return {"left": left, "top": top, "right": right, "bottom": bottom}

    def test_card_width_constant(self):
        """Card width must be 360px as specified."""
        # This mirrors the constant in main.js
        CARD_WIDTH = 360
        self.assertEqual(CARD_WIDTH, 360)

    def test_card_height_range(self):
        """Card height must be dynamic 150-240px with max 260px."""
        MIN_HEIGHT = 150
        MAX_HEIGHT = 260
        self.assertEqual(MIN_HEIGHT, 150)
        self.assertEqual(MAX_HEIGHT, 260)
        # Dynamic range 150-240 is within bounds
        self.assertGreaterEqual(MAX_HEIGHT, 240)
        self.assertLessEqual(MAX_HEIGHT, 260)

    def test_gap_between_12_and_16(self):
        """Gap between card bottom and float bar top must be 12-16px."""
        GAP = 14  # midpoint
        self.assertGreaterEqual(GAP, 12)
        self.assertLessEqual(GAP, 16)

    def test_position_above_float_bar_simple(self):
        """Card bottom edge must be above float bar top with GAP."""
        # Simulate: float bar at bottom of a 1080p display, centered
        float_bounds = {"x": 710, "y": 1046, "width": 500, "height": 34}
        display = self._make_display(0, 0, 1920, 1080)

        # Expected card position:
        # height = 200 (estimate), width = 360
        # card_bottom = float_top - GAP = 1046 - 14 = 1032
        # card_top = card_bottom - card_height = 1032 - 200 = 832
        # card_left = center on float bar: float_left + float_width/2 - card_width/2
        #           = 710 + 250 - 180 = 780
        card_height = 200
        card_width = 360
        gap = 14
        card_bottom = float_bounds["y"] - gap
        card_top = card_bottom - card_height
        card_left = float_bounds["x"] + (float_bounds["width"] // 2) - (card_width // 2)

        self.assertEqual(card_bottom, 1032)
        self.assertEqual(card_top, 832)
        self.assertEqual(card_left, 780)

    def test_position_uses_element_positions(self):
        """Position should use reported element positions, not full window bounds."""
        # The float window is 500x500 but visible bubble is at bottom-center
        element_pos = self._make_element_position(left=207, top=466, right=293, bottom=500)
        card_width = 360
        card_height = 200
        gap = 14

        # Horizontal center on element
        element_width = element_pos["right"] - element_pos["left"]
        expected_left = element_pos["left"] + (element_width // 2) - (card_width // 2)
        # Vertical: card bottom = element top - gap
        expected_top = element_pos["top"] - gap - card_height

        self.assertEqual(expected_left, 207 + 43 - 180)  # 70
        self.assertEqual(expected_top, 466 - 14 - 200)    # 252

    def test_position_clamps_to_workarea_left(self):
        """Card must not extend beyond left edge of workArea."""
        # Float bar near left edge
        float_bounds = {"x": 0, "y": 1000, "width": 500, "height": 34}
        display = self._make_display(0, 0, 1920, 1080)
        card_width = 360
        card_height = 200
        gap = 14

        card_left = float_bounds["x"] + (float_bounds["width"] // 2) - (card_width // 2)
        # Without clamping: 0 + 250 - 180 = 70 (already within bounds, no clamp needed)
        self.assertGreaterEqual(card_left, display.workArea["x"])

        # Extreme case: float bar at x=-100 (off-screen slightly)
        card_left = -100 + 250 - 180  # = -30
        clamped_left = max(card_left, display.workArea["x"])
        self.assertEqual(clamped_left, 0)

    def test_position_clamps_to_workarea_right(self):
        """Card must not extend beyond right edge of workArea."""
        display = self._make_display(0, 0, 1920, 1080)
        card_width = 360

        # Float bar near right edge: right edge of card would go past workArea right
        card_left = 1700  # 1700 + 360 = 2060 > 1920
        clamped_left = min(card_left, display.workArea["width"] - card_width)
        self.assertEqual(clamped_left, 1560)

    def test_position_clamps_to_workarea_top(self):
        """Card must not extend beyond top edge of workArea."""
        display = self._make_display(0, 0, 1920, 1080)

        # If card would go above workArea top, clamp it
        card_top = -20
        clamped_top = max(card_top, display.workArea["y"])
        self.assertEqual(clamped_top, 0)

    def test_position_clamps_to_workarea_bottom(self):
        """Card must not extend beyond bottom edge of workArea either."""
        display = self._make_display(0, 0, 1920, 1080)
        card_height = 260
        gap = 14

        # If float bar is very low, card bottom pushes past workArea bottom
        # card_bottom = float_top - gap = 1080 - 14 = 1066 (within bounds)
        # But if taskbar occupies bottom, workArea height < 1080
        display_small = self._make_display(0, 0, 1920, 1040)  # 40px taskbar
        float_bounds = {"x": 710, "y": 1006, "width": 500, "height": 34}

        card_bottom = float_bounds["y"] - gap
        display_bottom = display_small.workArea["y"] + display_small.workArea["height"]

        # Ensure card doesn't go below workArea
        self.assertLessEqual(card_bottom, display_bottom)

    def test_multi_display_current_display_used(self):
        """On multi-monitor, use currentDisplay instead of primary display."""
        # Two displays: primary 1920x1080 (x=0), secondary 1920x1080 (x=1920)
        primary = self._make_display(0, 0, 1920, 1080)
        secondary = self._make_display(1920, 0, 1920, 1080)

        # Float bar on secondary display
        float_bounds = {"x": 2630, "y": 1046, "width": 500, "height": 34}
        card_width = 360
        card_height = 200
        gap = 14

        # Card centered on float bar
        card_left = float_bounds["x"] + (float_bounds["width"] // 2) - (card_width // 2)
        card_top = float_bounds["y"] - gap - card_height

        # Must be within secondary display's workArea
        sec_work = secondary.workArea
        self.assertGreaterEqual(card_left, sec_work["x"])
        self.assertLess(card_left + card_width, sec_work["x"] + sec_work["width"])

    def test_fallback_to_floatwin_getbounds(self):
        """When elementPositions is empty, fall back to floatWin.getBounds()."""
        # Float window bounds: 500x500 centered at bottom of primary display
        float_bounds = {"x": 710, "y": 580, "width": 500, "height": 500}
        display = self._make_display(0, 0, 1920, 1080)

        # The visible bar is at the bottom-center of the float window
        # Estimated visible bar area: bottom 34px centered
        bar_height = 34
        bar_top = float_bounds["y"] + float_bounds["height"] - bar_height
        bar_width = 86  # from float.html: width=86
        bar_left = float_bounds["x"] + (float_bounds["width"] // 2) - (bar_width // 2)

        card_width = 360
        card_height = 200
        gap = 14

        # Position above estimated bar area
        card_left = bar_left + (bar_width // 2) - (card_width // 2)
        card_top = bar_top - gap - card_height

        # Basic sanity: card must be above the bar
        self.assertLess(card_top + card_height, bar_top)


if __name__ == "__main__":
    unittest.main()